server/sessionstore.go

```diff
@@ -104,19 +104,20 @@ func (ss *SessionStore) NewSession(conn interface{}, sid string) (*Session, int)
 		}
 	}
 
-	ss.lock.Unlock()
-
 	// Deleting long polling sessions.
 	for _, sess := range expired {
 		// This locks the session. Thus cleaning up outside of the
 		// sessionStore lock. Otherwise deadlock.
 		sess.cleanUp(true)
 	}
 
-	statsSet("LiveSessions", int64(len(ss.sessCache)))
+	numSessions := len(ss.sessCache)
+	statsSet("LiveSessions", int64(numSessions))
 	statsInc("TotalSessions", 1)
 
-	return &s, len(ss.sessCache)
+	ss.lock.Unlock()
```

Comment:
Why does it need to be moved here? If this is indeed needed, then there is no point in first collecting expired sessions to `expired ` and then `for _, sess := range expired`. You can call `sess.cleanUp(true)` as soon as the expired session is found. I explicitly wanted to expire session outside of the main lock to avoid locking sessions store for a potentially long time.

Reply:
The race detector complains about accessing `ss.sessCache` from multiple threads (`SessionStore.NewSession` is called from hdl_websock).
So any access to `sessCache` should be mutex protected:

Write at 0x00c0002e1530 by goroutine 83:
  runtime.mapassign_faststr()
      /usr/local/go/src/runtime/map_faststr.go:202 +0x0
  main.(*SessionStore).NewSession()
      tinode/chat/server/sessionstore.go:90 +0x721
  main.serveWebSocket()
      tinode/chat/server/hdl_websock.go:169 +0x7d8

Previous read at 0x00c0002e1530 by goroutine 82:
  main.(*SessionStore).NewSession()
      tinode/chat/server/sessionstore.go:119 +0xeb4
  main.serveWebSocket()
      tinode/chat/server/hdl_websock.go:169 +0x7d8

-----------------
I'm going to move `for _, sess := range expired` outside of the lock.

Reply:
Yes, that was my main concern: `for _, sess := range expired` does not need to be under the lock. `sess.cleanUp(true)` does not touch `sessionStore`.